### PR TITLE
Compiler Style

### PR DESCRIPTION
**Pull request by dns2utf8**
WIP PR

### CODE REVIEW

File: src/user_input_compiler.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
+//! # user_input_compiler
+//!
+//! Parses Input with this Grammer after trimming the Input and ignoring Whitespaces:
```
Comment (dbrgn) [inline src/user_input_compiler.rs:3]: Small detail: "Grammer" should be "grammar" and "Input" should be "input", see https://www.youtube.com/watch?v=N4vf8N6GpdM :grin:

File: src/user_input_compiler.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+//! 
+//! Command         := "/" CommandWord
+//! CommandWord     := Status | Subscribe | Cancel | InvalidSyntax
```
Comment (dbrgn) [inline src/user_input_compiler.rs]: Strictly speaking, a CommandWord is actually more than just a word. What about renaming Command to FullCommand and CommandWord to Command?

File: src/user_input_compiler.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+//! Subscribe       := "subscribe" SensorSelector Duration
+//! SensorSelector  := SensorString OptionalNum
+//! SensorString    := "account_balance" | "barometer" | "beverage_supply" | "door_locked" | "humidity" | "network_connections" | "power_consumption" | "temperature" | "total_member_count" | "radiation.alpha" | "radiation.beta_gamma" | "radiation.beta" | "radiation.gamma" | "people_now_present" | "wind"
+//! OptionalInteger := Integer | ɛ
+//! Integer         := [0-9]*
+//! Duration        := Real TimeSuffix
+//! TimeSuffix      := "m" | "min" | "h" | "d"
+//! Real            := Integer "." Integer | Integer
+//! Cancel          := "cancel"
+//! InvalidSyntax   := *
```
Comment (dbrgn) [inline src/user_input_compiler.rs:21]: I'm not sure whether parse errors should be part of the grammar. I think the lexer should be able to handle that.

File: src/user_input_compiler.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
+//! # user_input_compiler
+//!
+//! Parses Input with this Grammer after trimming the Input and ignoring Whitespaces:
+//! 
+//! Command         := "/" CommandWord
+//! CommandWord     := Status | Subscribe | Cancel | InvalidSyntax
+//! Status          := "status" | "crowd"
+//! Subscribe       := "subscribe" SensorSelector Duration
```
Comment (dbrgn) [inline src/user_input_compiler.rs:8]: Maybe we also need to specify a change threshold for some of the values (e.g. temperature).

**Comment by dbrgn** [comment]
Regarding the parser, if you don't want to use a parser generator I'd go with a simple recursive descent parser, where each line in the grammar is implemented by a function.

**Comment by dns2utf8** [comment]
Parser works :smile: 

Pro Tip: Never use ɛ in the middle of your Grammer

**Pull request merged**